src/phl/ecPhlOutline.ml

```diff
@@ -0,0 +1,409 @@
+    let tc = t_intros_i_1 [pl; pr; h] tc in
+    let goal = tc1_goal tc in
+
+    let p = match side with | `Left -> pl | `Right -> pr in
+    let b = match side with | `Left -> true | `Right -> false in
+
+    let handle_exists () =
+      (* Pairing up the correct variables for the exists intro *)
+      let vs, fm = EcFol.destr_exists goal in
+      let eqs_pre, _ =
+        let l, r = EcFol.destr_and fm in
+        if b then l, r else r, l
+      in
+      let eqs, _ = destr_and eqs_pre in
+      let eqs = destr_ands ~deep:false eqs in
+      let doit eq =
+        let l, r = destr_eq eq in
+        let l, r = if b then r, l else l, r in
+        let v = EcFol.destr_local l in
+        v, r
+      in
+      let eqs = List.map doit eqs in
+      let exvs =
+        List.map
+          (fun (id, _) ->
+            let v = List.assoc id eqs in
+            Fsubst.f_subst_mem (EcMemory.memory m) p v)
+          vs
+      in
+
+      as_tcenv1 (t_exists_intro_s (List.map paformula exvs) tc)
+    in
+
+    let tc =
+      if EcFol.is_exists goal then
+        handle_exists ()
+      else
+        tc
+    in
+
+    t_seq
+      (t_generalize_hyp ?clear:(Some `Yes) h)
+      EcHiGoal.process_done
+      tc
+  in
+
+  let tc = t_seqsub
+             (EcPhlTrans.t_equivS_trans (EcMemory.memtype m, s) c1 c2)
+             [exists_subtac; EcHiGoal.process_done; t_id; t_id]
+             tc
+  in
+  tc
+
+(*---------------------------------------------------------------------------------------*)
+(*
+   This is the improved transitivity star from above but allows for a range of code
+   positions to select the program slice that is changing. The prefix and suffix are
+   extracted and copied across to build the new program.
+*)
+(* FIXME: Maybe move to ecPhlTrans as well *)
+
+let t_outline_stmt side start_pos end_pos s tc =
+  let goal = tc1_as_equivS tc in
+
+  (* Check which memory/program we are outlining *)
+  let code = match side with
+    | `Left  -> goal.es_sl
+    | `Right -> goal.es_sr
+  in
+
+  (* Extract the program prefix and suffix *)
+  let rest, code_suff  = s_split end_pos code in
+  let code_pref, _, _ = s_split_i start_pos (stmt rest) in
+
+  let new_prog = s_seq (s_seq (stmt code_pref) s) (stmt code_suff) in
+  let tc = t_equivS_trans_eq side new_prog tc in
+
+  (* The middle goal, showing equivalence with the replaced code, ideally solves. *)
+  let tp = match side with | `Left -> 1 | `Right -> 2 in
+  let p = EcHiGoal.process_tfocus tc (Some [Some tp, Some tp], None) in
+  let tc =
+    t_onselect
+      p
+      (t_try (
+           t_seqs [
+               EcPhlInline.process_inline (`ByName (None, None, ([], None)));
+               EcPhlEqobs.process_eqobs_in none {sim_pos = none; sim_hint = ([], none); sim_eqs = none};
+               EcPhlAuto.t_auto;
+               EcHiGoal.process_done;
+         ]))
+      tc
+  in
+  tc
+
+(*---------------------------------------------------------------------------------------*)
```

Comment:
Does this unification procedure is general enough to be factored out of this file?